File: backend/app/schemas/staff.py
from __future__ import annotations
import uuid
from datetime import date, datetime
from pydantic import BaseModel, field_validator
from app.models.staff import EmploymentType, Gender, MaritalStatus, StaffType
from app.models.staff_history import LeaveStatus


class StaffCategoryCreate(BaseModel):
    name: str
    code: str
    staff_type: StaffType | None = None

    @field_validator("name", "code")
    @classmethod
    def not_blank(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Field must not be blank")
        return v.strip()


class StaffCategoryRead(BaseModel):
    id: uuid.UUID
    name: str
    code: str
    staff_type: StaffType | None
    is_template: bool
    is_active: bool
    school_id: uuid.UUID | None
    model_config = {"from_attributes": True}


class StaffRankCreate(BaseModel):
    title: str
    category_id: uuid.UUID

    @field_validator("title")
    @classmethod
    def not_blank(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Field must not be blank")
        return v.strip()


class StaffRankRead(BaseModel):
    id: uuid.UUID
    title: str
    category_id: uuid.UUID
    is_template: bool
    is_active: bool
    school_id: uuid.UUID | None
    model_config = {"from_attributes": True}


class PositionRead(BaseModel):
    id: uuid.UUID
    name: str
    model_config = {"from_attributes": True}


class StaffMemberCreate(BaseModel):
    staff_number: str
    first_name: str
    middle_name: str | None = None
    last_name: str
    category_id: uuid.UUID | None = None
    date_of_birth: date | None = None
    gender: Gender | None = None
    employment_type: EmploymentType | None = None
    marital_status: MaritalStatus | None = None
    national_id: str | None = None
    ssnit_number: str | None = None
    address: str | None = None
    phone: str | None = None
    email: str | None = None
    position_ids: list[uuid.UUID] = []
    joined_date: date | None = None

    @field_validator("staff_number", "first_name", "last_name")
    @classmethod
    def not_blank(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Field must not be blank")
        return v.strip()


class StaffMemberUpdate(BaseModel):
    first_name: str | None = None
    middle_name: str | None = None
    last_name: str | None = None
    category_id: uuid.UUID | None = None
    date_of_birth: date | None = None
    gender: Gender | None = None
    employment_type: EmploymentType | None = None
    marital_status: MaritalStatus | None = None
    national_id: str | None = None
    ssnit_number: str | None = None
    address: str | None = None
    phone: str | None = None
    email: str | None = None
    position_ids: list[uuid.UUID] | None = None  # None = no change; [] = clear all
    is_active: bool | None = None
    joined_date: date | None = None


class StaffMemberSummary(BaseModel):
    """Flat view used in list responses — no nested collections."""
    id: uuid.UUID
    school_id: uuid.UUID
    staff_number: str
    first_name: str
    middle_name: str | None
    last_name: str
    display_name: str
    category_id: uuid.UUID | None
    category_name: str | None
    staff_type: StaffType | None
    gender: Gender | None
    employment_type: EmploymentType | None
    phone: str | None
    email: str | None
    position_ids: list[uuid.UUID]
    position_names: list[str]
    is_active: bool
    joined_date: date | None
    photo_url: str | None = None


class StaffMemberDetail(StaffMemberSummary):
    """Full profile view — includes nested sub-records."""
    date_of_birth: date | None
    marital_status: MaritalStatus | None
    national_id: str | None
    ssnit_number: str | None
    address: str | None
    photo_path: str | None
    has_account: bool
    qualifications: list[QualificationRead] = []
    emergency_contacts: list[EmergencyContactRead] = []


class EmergencyContactCreate(BaseModel):
    name: str
    contact_type: str    # "Spouse", "Parent", "Sibling", etc.
    phone: str
    email: str | None = None


class EmergencyContactRead(BaseModel):
    id: uuid.UUID
    name: str
    contact_type: str
    phone: str
    email: str | None

    model_config = {"from_attributes": True}


class QualificationCreate(BaseModel):
    institution: str
    qualification_type: str    # "Bachelor's", "Master's", "PGCE", etc.
    field_of_study: str | None = None
    year_obtained: int | None = None


class QualificationUpdate(BaseModel):
    institution: str | None = None
    qualification_type: str | None = None
    field_of_study: str | None = None
    year_obtained: int | None = None


class QualificationRead(BaseModel):
    id: uuid.UUID
    institution: str
    qualification_type: str
    field_of_study: str | None
    year_obtained: int | None

    model_config = {"from_attributes": True}


class PromotionCreate(BaseModel):
    from_rank_id: uuid.UUID | None = None
    to_rank_id: uuid.UUID
    effective_date: date
    reason: str | None = None


class PromotionUpdate(BaseModel):
    from_rank_id: uuid.UUID | None = None
    to_rank_id: uuid.UUID | None = None
    effective_date: date | None = None
    reason: str | None = None


class PromotionRead(BaseModel):
    id: uuid.UUID
    staff_member_id: uuid.UUID
    from_rank_id: uuid.UUID | None
    to_rank_id: uuid.UUID | None
    from_rank_title: str | None = None
    to_rank_title: str | None = None
    effective_date: date
    reason: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class LeaveCreate(BaseModel):
    leave_type: str    # "Annual", "Sick", "Maternity", "Study", etc.
    start_date: date
    end_date: date
    # Caller must supply days_count: the school calendar may exclude weekends
    # and public holidays, so the server cannot compute it reliably here.
    days_count: int
    reason: str | None = None

    @field_validator("days_count")
    @classmethod
    def positive_days(cls, v: int) -> int:
        if v < 1:
            raise ValueError("days_count must be at least 1")
        return v


class LeaveRead(BaseModel):
    id: uuid.UUID
    staff_member_id: uuid.UUID
    staff_name: str | None = None
    staff_number: str | None = None
    leave_type: str
    start_date: date
    end_date: date
    days_count: int
    reason: str | None
    status: LeaveStatus
    approved_by_id: uuid.UUID | None
    reviewed_at: datetime | None
    notes: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class LeaveReview(BaseModel):
    """Body for the approve/reject endpoint."""
    status: LeaveStatus
    notes: str | None = None

    @field_validator("status")
    @classmethod
    def terminal_status_only(cls, v: LeaveStatus) -> LeaveStatus:
        if v not in (LeaveStatus.APPROVED, LeaveStatus.REJECTED):
            raise ValueError("status must be APPROVED or REJECTED")
        return v

# Resolve forward references for StaffMemberDetail
StaffMemberDetail.model_rebuild()

class TempPasswordResult(BaseModel):
    temporary_password: str
    display_name: str


# ── Responsibilities ──────────────────────────────────────────────────────────

class ClassTeacherAssignment(BaseModel):
    class_id: uuid.UUID
    class_name: str
    academic_year_id: uuid.UUID
    academic_year_name: str
    is_active: bool


class SubjectAssignment(BaseModel):
    class_id: uuid.UUID
    class_name: str
    subject_id: uuid.UUID
    subject_name: str
    academic_year_id: uuid.UUID
    academic_year_name: str
    is_active: bool


class HouseAssignment(BaseModel):
    house_id: uuid.UUID
    house_name: str
    house_code: str
    academic_year_id: uuid.UUID
    academic_year_name: str
    is_active: bool


class StaffResponsibilities(BaseModel):
    class_teacher: ClassTeacherAssignment | None
    subject_assignments: list[SubjectAssignment]
    house_assignments: list[HouseAssignment]

File: frontend/src/lib/api/staff.ts
import client from './client';

export type Gender = 'MALE' | 'FEMALE';
export type EmploymentType = 'PERMANENT' | 'CONTRACT' | 'NATIONAL_SERVICE' | 'INTERN';
export type MaritalStatus = 'SINGLE' | 'MARRIED' | 'DIVORCED' | 'WIDOWED' | 'SEPARATED';
export type StaffType = 'TEACHING' | 'NON_TEACHING';

export interface StaffCategory {
  id: string;
  name: string;
  code: string;
  staff_type: StaffType | null;
  is_template: boolean;
  is_active: boolean;
  school_id: string | null;
}

export interface StaffRank {
  id: string;
  title: string;
  category_id: string;
  is_template: boolean;
  is_active: boolean;
  school_id: string | null;
}

export interface StaffSummary {
  id: string;
  school_id: string;
  staff_number: string;
  first_name: string;
  middle_name: string | null;
  last_name: string;
  display_name: string;
  category_id: string | null;
  category_name: string | null;
  staff_type: StaffType | null;
  gender: Gender | null;
  employment_type: EmploymentType | null;
  phone: string | null;
  email: string | null;
  position_ids: string[];
  position_names: string[];
  is_active: boolean;
  joined_date: string | null;
}

export interface Qualification {
  id: string;
  institution: string;
  qualification_type: string;
  field_of_study: string | null;
  year_obtained: number | null;
}

export interface EmergencyContact {
  id: string;
  name: string;
  contact_type: string;
  phone: string;
  email: string | null;
}

export interface StaffDetail extends StaffSummary {
  date_of_birth: string | null;
  marital_status: MaritalStatus | null;
  national_id: string | null;
  ssnit_number: string | null;
  address: string | null;
  photo_path: string | null;
  has_account: boolean;
  qualifications: Qualification[];
  emergency_contacts: EmergencyContact[];
}

export interface Position {
  id: string;
  name: string;
}

export interface Promotion {
  id: string;
  staff_member_id: string;
  from_rank_id: string | null;
  to_rank_id: string | null;
  from_rank_title: string | null;
  to_rank_title: string | null;
  effective_date: string;
  reason: string | null;
  created_at: string;
}

export type LeaveStatus = 'PENDING' | 'APPROVED' | 'REJECTED' | 'CANCELLED';

export interface Leave {
  id: string;
  staff_member_id: string;
  staff_name: string | null;
  staff_number: string | null;
  leave_type: string;
  start_date: string;
  end_date: string;
  days_count: number;
  reason: string | null;
  status: LeaveStatus;
  approved_by_id: string | null;
  reviewed_at: string | null;
  notes: string | null;
  created_at: string;
}

export interface StaffListParams {
  active_only?: boolean;
  skip?: number;
  limit?: number;
  search?: string;
  gender?: string;
  category_id?: string;
}

export async function listStaff(params?: StaffListParams): Promise<StaffSummary[]> {
  const { data } = await client.get<StaffSummary[]>('/staff', { params });
  return data;
}

export interface StaffListPage {
  items: StaffSummary[];
  total: number;
}

export async function listStaffPage(params: StaffListParams = {}): Promise<StaffListPage> {
  const res = await client.get<StaffSummary[]>('/staff', { params });
  return { items: res.data, total: Number(res.headers['x-total-count'] ?? res.data.length) };
}

export async function getStaff(id: string): Promise<StaffDetail> {
  const { data } = await client.get<StaffDetail>(`/staff/${id}`);
  return data;
}

export async function createStaff(req: {
  staff_number: string;
  first_name: string;
  middle_name?: string;
  last_name: string;
  category_id?: string;
  date_of_birth?: string;
  gender?: Gender;
  employment_type?: EmploymentType;
  marital_status?: MaritalStatus;
  national_id?: string;
  ssnit_number?: string;
  address?: string;
  phone?: string;
  email?: string;
  position_ids?: string[];
  joined_date?: string;
}): Promise<StaffDetail> {
  const { data } = await client.post<StaffDetail>('/staff', req);
  return data;
}

export async function updateStaff(
  id: string,
  req: Partial<{
    first_name: string;
    middle_name: string;
    last_name: string;
    category_id: string;
    date_of_birth: string;
    joined_date: string;
    gender: Gender;
    employment_type: EmploymentType;
    marital_status: MaritalStatus;
    national_id: string;
    ssnit_number: string;
    address: string;
    phone: string;
    email: string;
    position_ids: string[];
    is_active: boolean;
  }>
): Promise<StaffDetail> {
  const { data } = await client.patch<StaffDetail>(`/staff/${id}`, req);
  return data;
}

export async function addQualification(
  staffId: string,
  req: {
    institution: string;
    qualification_type: string;
    field_of_study?: string;
    year_obtained?: number;
  }
): Promise<Qualification> {
  const { data } = await client.post<Qualification>(`/staff/${staffId}/qualifications`, req);
  return data;
}

export async function updateQualification(
  staffId: string,
  qualId: string,
  req: {
    institution?: string;
    qualification_type?: string;
    field_of_study?: string;
    year_obtained?: number;
  }
): Promise<Qualification> {
  const { data } = await client.patch<Qualification>(`/staff/${staffId}/qualifications/${qualId}`, req);
  return data;
}

export async function deleteQualification(staffId: string, qualId: string): Promise<void> {
  await client.delete(`/staff/${staffId}/qualifications/${qualId}`);
}

export async function addEmergencyContact(
  staffId: string,
  req: { name: string; contact_type: string; phone: string; email?: string }
): Promise<EmergencyContact> {
  const { data } = await client.post<EmergencyContact>(
    `/staff/${staffId}/emergency-contacts`,
    req
  );
  return data;
}

export async function deleteEmergencyContact(staffId: string, contactId: string): Promise<void> {
  await client.delete(`/staff/${staffId}/emergency-contacts/${contactId}`);
}

export async function listPositions(): Promise<Position[]> {
  const { data } = await client.get<Position[]>('/schools/me/positions');
  return data;
}

export async function listCategories(): Promise<StaffCategory[]> {
  const { data } = await client.get<StaffCategory[]>('/staff/categories');
  return data;
}

export async function listRanks(categoryId: string): Promise<StaffRank[]> {
  const { data } = await client.get<StaffRank[]>('/staff/ranks', { params: { category_id: categoryId } });
  return data;
}

export async function listPromotions(staffId: string): Promise<Promotion[]> {
  const { data } = await client.get<Promotion[]>(`/staff/${staffId}/promotions`);
  return data;
}

export async function addPromotion(
  staffId: string,
  req: {
    from_rank_id?: string;
    to_rank_id: string;
    effective_date: string;
    reason?: string;
  }
): Promise<Promotion> {
  const { data } = await client.post<Promotion>(`/staff/${staffId}/promotions`, req);
  return data;
}

export async function updatePromotion(
  staffId: string,
  promotionId: string,
  req: {
    from_rank_id?: string;
    to_rank_id?: string;
    effective_date?: string;
    reason?: string;
  }
): Promise<Promotion> {
  const { data } = await client.patch<Promotion>(`/staff/${staffId}/promotions/${promotionId}`, req);
  return data;
}

export async function deletePromotion(staffId: string, promotionId: string): Promise<void> {
  await client.delete(`/staff/${staffId}/promotions/${promotionId}`);
}

export interface StaffExportParams {
  category_id?: string;
  active_only?: boolean;
  search?: string;
  gender?: string;
}

export async function exportStaffExcel(params?: StaffExportParams): Promise<void> {
  const response = await client.get('/staff/export/excel', { params, responseType: 'blob' });
  const url = URL.createObjectURL(new Blob([response.data]));
  const a = Object.assign(document.createElement('a'), { href: url, download: 'staff_register.xlsx' });
  a.click();
  URL.revokeObjectURL(url);
}

export interface StaffCustomExportParams extends StaffExportParams {
  fields: string;
  fmt: 'csv' | 'excel';
}

export async function customExportStaff(params: StaffCustomExportParams): Promise<void> {
  const response = await client.get('/staff/export/custom', { params, responseType: 'blob' });
  const ext = params.fmt === 'excel' ? 'xlsx' : 'csv';
  const url = URL.createObjectURL(new Blob([response.data]));
  const a = Object.assign(document.createElement('a'), { href: url, download: `staff.${ext}` });
  a.click();
  URL.revokeObjectURL(url);
}

export async function exportStaffPdf(params?: StaffExportParams): Promise<void> {
  const response = await client.get('/staff/export/pdf', { params, responseType: 'blob' });
  const url = URL.createObjectURL(new Blob([response.data]));
  const a = Object.assign(document.createElement('a'), { href: url, download: 'staff_register.pdf' });
  a.click();
  URL.revokeObjectURL(url);
}

export async function listLeave(staffId: string): Promise<Leave[]> {
  const { data } = await client.get<Leave[]>(`/staff/${staffId}/leave`);
  return data;
}

export async function submitLeave(
  staffId: string,
  req: {
    leave_type: string;
    start_date: string;
    end_date: string;
    days_count: number;
    reason?: string;
  }
): Promise<Leave> {
  const { data } = await client.post<Leave>(`/staff/${staffId}/leave`, req);
  return data;
}

export async function listPendingLeave(): Promise<Leave[]> {
  const { data } = await client.get<Leave[]>('/staff/leave/pending');
  return data;
}

export async function reviewLeave(
  leaveId: string,
  req: { status: 'APPROVED' | 'REJECTED'; notes?: string }
): Promise<Leave> {
  const { data } = await client.patch<Leave>(`/staff/leave/${leaveId}/review`, req);
  return data;
}

export interface ImportRowResult {
  row: number;
  ref: string | null;
  status: string;
  error: string | null;
  warning: string | null;
}

export interface ImportBatchResult {
  batch_id: string;
  total_rows: number;
  created: number;
  failed: number;
  errors: ImportRowResult[];
  warnings: ImportRowResult[];
}

export async function downloadImportTemplate(): Promise<void> {
  const response = await client.get('/staff/import/template', { responseType: 'blob' });
  const cd = response.headers['content-disposition'] as string | undefined;
  const match = cd?.match(/filename="?([^";]+)"?/);
  const filename = match?.[1] ?? 'staff_import_template.xlsx';
  const url = URL.createObjectURL(new Blob([response.data]));
  const a = Object.assign(document.createElement('a'), { href: url, download: filename });
  a.click();
  URL.revokeObjectURL(url);
}

export async function bulkImportStaff(file: File): Promise<ImportBatchResult> {
  const form = new FormData();
  form.append('file', file);
  const { data } = await client.post<ImportBatchResult>('/staff/import', form);
  return data;
}

export interface TempPasswordResult {
  temporary_password: string;
  display_name: string;
}

export async function resetStaffPassword(staffId: string): Promise<TempPasswordResult> {
  const { data } = await client.post<TempPasswordResult>(`/staff/${staffId}/reset-password`);
  return data;
}

export interface InviteResult {
  invitation_token: string;
  invite_link: string;
  sms_sent: boolean;
}

export async function inviteStaff(staffId: string): Promise<InviteResult> {
  const { data } = await client.post<InviteResult>(`/staff/${staffId}/invite`);
  return data;
}

export interface ClassTeacherAssignment {
  class_id: string;
  class_name: string;
  academic_year_id: string;
  academic_year_name: string;
  is_active: boolean;
}

export interface SubjectAssignment {
  class_id: string;
  class_name: string;
  subject_id: string;
  subject_name: string;
  academic_term_id: string;
  term_name: string;
  academic_year_name: string;
  is_active: boolean;
}

export interface HouseAssignment {
  house_id: string;
  house_name: string;
  house_code: string;
  academic_year_id: string;
  academic_year_name: string;
  is_active: boolean;
}

export interface StaffResponsibilities {
  class_teacher: ClassTeacherAssignment | null;
  subject_assignments: SubjectAssignment[];
  house_assignments: HouseAssignment[];
}

export async function getStaffResponsibilities(staffId: string): Promise<StaffResponsibilities> {
  const { data } = await client.get<StaffResponsibilities>(`/staff/${staffId}/responsibilities`);
  return data;
}

export type PermissionSource = 'override' | 'position' | 'default_deny';

export interface StaffPermission {
  module: string;
  action: string;
  effective: boolean;
  source: PermissionSource;
  override_is_allowed: boolean | null;
}

export const listStaffPermissions = (staffId: string): Promise<StaffPermission[]> =>
  client.get<StaffPermission[]>(`/staff/${staffId}/permissions`).then(r => r.data);

export const setStaffPermission = (
  staffId: string,
  module: string,
  action: string,
  is_allowed: boolean,
): Promise<StaffPermission[]> =>
  client.post<StaffPermission[]>(`/staff/${staffId}/permissions`, { module, action, is_allowed }).then(r => r.data);

export const clearStaffPermission = (
  staffId: string,
  module: string,
  action: string,
): Promise<StaffPermission[]> =>
  client.delete<StaffPermission[]>(`/staff/${staffId}/permissions/${module}/${action}`).then(r => r.data);
